### PR TITLE
Implemented struct-based enumerator

### PR DESCRIPTION
**Pull request by RaZeR-RBI**
First of all, thank you very much for making this library.

I'm currently using it in an environment that is sensitive to heap allocations (running in almost-realtime) and noticed that the enumerator exposed by this library is implemented as an iterator method (using **yield**), which causes heap allocations.

I've implemented a struct-based enumerator that should provide the same guarantees as the original one. The original one is exposed as an explicit IEnumerable<T> implementation.

If any improvements could be made or I have an issue in my implementation (I've tested it in my current project though), let me know.

### CODE REVIEW

**Comment by i3arnon** [comment]
Thank you for the kind words.

I've been thinking about this change quite a lot.
I always try to keep the code here as close as possible to .NET's `ConcurrentDictionary`.
It obviously has the same issues with allocations but they haven't yet offered a public struct enumerator due to binary backwards compatibility (see [here](https://github.com/dotnet/runtime/issues/25448))
What they however did do in the meantime is implement a private struct enumerator which is resettable, so you would be able to obtain it once (with a boxing allocation) and reset it and reuse (see [here](https://github.com/dotnet/runtime/pull/38296)).

I think not breaking binary back compat is too high of a bar for this library so I'm willing to go further and change the enumerator (it should still keep source back compat).
But I still would like the implementation to be as close as possible to [.NET's](https://github.com/dotnet/runtime/blob/ca33c4dbb57ea8be14fdaeae00364aee2ede5ba3/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L811-L891)

Are you willing to close that gap?

A few notes right off the bat:
1. When offering a struct enumerator, the overload necessary for implementing `IEnumerable` should still return the **boxed** struct enumerator. You don't want to maintain different implementations.
1. While I get throwing instead of returning an `object` in order to avoid boxing (I do that in my code to make sure I don't box accidently), I can't do that in a publicly used library.

**Comment by RaZeR-RBI** [comment]
Thanks for the input! I've updated the implementation.
The non-generic IEnumerator implementation shouldn't have been throwing, that's definitely my oversight. I've replaced it with a proper implementation, though I don't 100% like it because for some reason I couldn't put up an elvis operator here to shorten things up (something like ```current?.Item ?? null``` instead of a ternary). Perhaps I'm missing something obvious.

I've also updated IEnumerator<T>.Current to return default(T) instead of throwing, which should be more in line with what the .NET does.

---

Now about the ConcurrentDictionary's enumerator.
I totally understand the reasoning behind their implementation since it's a base class library and it's expected to be as compatible as possible (and they do a great job at keeping it that way), but I do agree that since we're not providing the same guarantees it's probably ok to go further and break the binary compatibility while saving the source/API one (which, at least for me, is more of a concern when upgrading my dependencies).

We could probably do the same as them and replace the struct with a resettable class (it's marked as class in the linked impl), but I don't particularly like it for the following reasons:
- To avoid additional allocations some sort of pooling or caching is needed; we also need to ensure that when the functionality that uses the enumerator gets out of scope (gets collected by GC) the enumerators should be evicted too to allow old references to buckets to be garbage-collected. It's quite unwieldy in my opinion and is probably hard to do correctly in complex scenarios.
- AFAIK iterating over an enumerator directly is not supported by **foreach**, so in order to reuse the enumerator we should use the **while** loop instead which is a bit less readable (though I may be wrong here).

To sum it up, I'm totally for changing the enumerator (but I'm obviously biased towards it since I'm making the PR for that particular case :smile:) because it would be way more ergonomic IMO.

But I think we should borrow their idea of encapsulating the collection's class instance in order to get fresh buckets array after each Reset() so people wouldn't get stale data in case they decide to consume it after the tables get resized. What do you think?

P.S.: Is there any test suite that's tied to this library? I understand that it's hard to test for concurrency issues, but at least some sanity checks would be nice. I'm a bit afraid of making changes without having at least some sort of verification besides 'it works on my machine' :wink:

**Comment by i3arnon** [comment]
> What do you think?

I think going with their implementation, changing it to fit `ConcurrentHashSet` and making it a struct instead of class is the way to go.
Even though their "state machine" style implementation is needed for a ressetable class implementation and less so for a struct one.
Keeping the codebases close allows to keep absorbing newer changes and optimizations.



> Is there any test suite that's tied to this library?

No 😄  Which is one of the reasons I want to make sure the code is close to its origins.
I'm open for contributions though 😉

**Comment by RaZeR-RBI** [comment]
> I think going with their implementation, changing it to fit `ConcurrentHashSet` and making it a struct instead of class is the way to go.
> Even though their "state machine" style implementation is needed for a ressetable class implementation and less so for a struct one.
> Keeping the codebases close allows to keep absorbing newer changes and optimizations.

Sounds good. I've ported *(shamelessly copied with minor edits)* their implementation.


> No :smile: Which is one of the reasons I want to make sure the code is close to its origins.
> I'm open for contributions though :wink:

That's risky. Well, in the end it doesn't matter much if it works though :wink:
Probably we should file a separate issue for this and mark it as `up for grabs` or something in case someone will decide to help us out.

**Comment by i3arnon** [comment]
Thank you @RaZeR-RBI.
The new version is published here: https://www.nuget.org/packages/ConcurrentHashSet/1.3.0
